submissions/expero1/oop-exercise/index.js

```diff
@@ -0,0 +1,105 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: _put repo URL here_
+   Web app: _put project's github pages URL here_
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+// Define your objects here
+
+class Habitat {
+  static SPECIES = { DOG: "dog", CAT: "cat", HUMAN: "human" };
+  static GENDER = { MALE: "male", FEMALE: "female" };
+  friends = [];
+  PROPS = ["species", "name", "gender", "legs", "saying"];
+  constructor(species, name, gender, legs, saying) {
+    this.species = species;
+    this.name = name;
+    this.gender = gender;
+    this.legs = legs;
+    this.saying = saying;
+  }
+
+  addFriend(habitat) {
+    if (!this.friends.includes(habitat)) {
+      this.friends.push(habitat);
+    }
+    if (!habitat.friends.includes(this)) {
+      habitat.friends.push(this);
+    }
+  }
+
+  toString() {
+    let propsKeys = this.PROPS.map((key) => {
+      return this[key];
+    });
+    let info = propsKeys.join("; ");
+    if (this.friends.length > 0) {
+      info += "; ";
+      let friendList = this.friends.reduce((list, friend) => {
+        list.push(friend.species + " " + friend.name);
+        return list;
+      }, []);
+
+      info += friendList.join(", ");
+    }
+    return info;
+  }
+}
+
+class Human extends Habitat {
+  constructor(name, gender, legs, hands, saying) {
+    super(Habitat.SPECIES.HUMAN, name, gender, legs, saying);
+    this.hands = hands;
+    this.PROPS.splice(this.PROPS.indexOf('legs'), 0, 'hands')
```

Comment:
Child classes shouldn't rely on base class internals. Imagine, the hierarchy is big, split across files and someone is tasked to rename `legs` on the base class into something else. It will take time to find a bug.
Root class could provide mechanics to group any types of limbs, since we want them to stay close.
This would probably lead to treat compound object properties as a special case, just like friends that are a special case of array property.

Reply:
Thanks for the important note. I tried to do as you recommended. I hope it worked out exactly as you said. This is a very important experience for me in writing code. Since it is one thing to know the basics of a programming language. And it’s a completely different thing when you have to apply certain patterns and avoid typical mistakes that are now invisible, but in the future can lead to errors that will take a long time to look for. Thanks again, with your help I've gotten a little better at coding.